CPT4/load_stage.sql

```diff
@@ -48,36 +46,66 @@ INSERT INTO concept_stage (
-	NULL AS domain_id, -- adding manually
-	'CPT4' AS vocabulary_id,
+SELECT DISTINCT vocabulary_pack.CutConceptName(UPPER(SUBSTRING(str FROM 1 FOR 1)) ||substring(str FROM 2 FOR LENGTH(str))) AS concept_name,   -- field with a term name from mrconso
+	'' AS domain_id, -- is about to be assigned at the end
+	'CPT4' AS vocabulary_id, 
 	'CPT4' AS concept_class_id,
 	'S' AS standard_concept,
-	scui AS concept_code,
-	(
-		SELECT latest_update
-		FROM vocabulary
-		WHERE vocabulary_id = 'CPT4'
-		) AS valid_start_date,
+	scui AS concept_code, -- = mrconso.code
+	(SELECT latest_update
+	FROM vocabulary
+	WHERE vocabulary_id = 'CPT4'
+	) AS valid_start_date,
 	TO_DATE('20991231', 'yyyymmdd') AS valid_end_date,
 	NULL AS invalid_reason
 FROM sources.mrconso
+   JOIN umls_mrsty USING (cui) -- need to add this table to sources https://download.nlm.nih.gov/umls/kss/2020AA/umls-2020AA-metathesaurus.zip => MRSTY
 WHERE sab = 'CPT'
 	AND suppress NOT IN (
-		'E',
-		'O',
-		'Y'
+		'E', -- Non-obsolete content marked suppressible by an editor
+		'O', -- All obsolete content, whether they are obsolesced by the source or by NLM
+		'Y' -- Non-obsolete content deemed suppressible during inversion
 		)
 	AND tty IN (
-		'PT',
-		'GLP'
+		'PT', -- Designated preferred name
+		'GLP' -- Global period
 		)
+; -- 10488
 
-UNION ALL
-
-VALUES ('Infectious agent detection by nucleic acid (DNA or RNA); severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) (Coronavirus disease [COVID-19]), amplified probe technique','Measurement','CPT4','CPT4','S','87635',TO_DATE('20200313','yyyymmdd'),TO_DATE('20991231','yyyymmdd'), NULL);
+--4. Add CPT4 codes which have no entry in sab = 'CPT' (only sab = 'HCPT'). Note, they are not HCPCS codes! 
+INSERT INTO concept_stage (
+	concept_name,
+	domain_id,
+	vocabulary_id,
+	concept_class_id,
+	standard_concept,
+	concept_code,
+	valid_start_date,
+	valid_end_date,
+	invalid_reason
+	)
+SELECT vocabulary_pack.CutConceptName(UPPER(SUBSTRING(str FROM 1 FOR 1)) ||substring(str FROM 2 FOR LENGTH(str))) AS concept_name,
+        '' AS domain_id, -- is about to be assigned at the end
+       'CPT4' AS vocabulary_id,
+       'CPT4' AS concept_class_id,
+       'S' AS standard_concept,
+       scui AS concept_code,
+       (SELECT latest_update
+        	FROM vocabulary
+        	WHERE vocabulary_id = 'CPT4') AS valid_start_date,
+       TO_DATE('20991231','yyyymmdd') AS valid_end_date,
+       NULL as invalid_reason
+  FROM sources.mrconso
+  JOIN umls_mrsty USING (cui)
+  WHERE scui IN (SELECT scui FROM sources.mrconso WHERE sab = 'HCPT')
+  AND   scui NOT IN (SELECT scui FROM sources.mrconso WHERE sab = 'CPT')
+  AND ((tty = 'PT'
+  AND suppress = 'N') or (tty = 'OP' and suppress = 'O'))
+AND scui NOT IN (SELECT concept_code
+FROM concept
+WHERE vocabulary_id = 'CPT4'); -- 29
 
--- Place of Sevice (POS) CPT terms
+--5. Add Place of Sevice (POS) CPT terms which do not appear in patient data and used for hierarchical search
```

Comment:
you define domains in steps 15,16.
Why not to add POS to step 3 and then define domain for them in step 15?

Reply:
I thought about it but did not find enough weighty reasons.  Maybe your approach is more elegant. Can re-do